[DOC-12610]: document `nprobe` tweaking

RayOffiah opened the pull request:
New page on fine-tuning the vector search.

RayOffiah committed:
[quickfix]: quickfix: strange font problem on diagram (#274)

Seems to have been caused by extra space in the label.

RayOffiah committed:
[DOC-12610]: Document `nprobe` tweaking

sarahlwelton commented on modules/tutorials/partials/diagrams/couchbase-hierarchy.puml:
Did this need to be included in this PR?

RayOffiah replied:
I'll move it to a separate fix.

sarahlwelton commented on modules/vector-search/partials/nav.adoc:
```diff
@@ -1,6 +1,7 @@
 * xref:7.6@server:vector-search:vector-search.adoc[]
 ** xref:7.6@server:vector-search:create-vector-search-index-ui.adoc[]
 ** xref:7.6@server:vector-search:create-vector-search-index-rest-api.adoc[]
+** xref:7.6@server:vector-search:fine-tune-vector-search.adoc[]
```
I would move this below all of the Run a Vector Search pages in the TOC, personally.

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+= Fine-tuning Vector Searches
```
```suggestion
= Fine-Tuning a Vector Search Query
```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+:page-ui-name: {ui-name}
+:page-product-name: {product-name}
+:stem:
+:description: pass:q[Additional parameters can be added to a vector search REST call that will allow you to tune the search for recall and/or accuracy.]
```
```suggestion
:description: Add additional parameters to a Vector Search REST API call to tune the search for recall or accuracy.
```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+= Fine-tuning Vector Searches
+:page-topic-type: concept
```
I'm not sure that this counts as a conceptual page.

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+{description}
+
+[#overview]
+== Overview
```
If you're not linking to this from elsewhere, this isn't a useful heading. 

```suggestion

```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+[#overview]
+== Overview
+
+Couchbase automatically dynamically tunes vector indexes to achieve a balance between recall (the quality of the search results),
+latency (search response time), and memory efficiency.
+This tuning occurs during indexing and querying, eliminating the need for users to adjust optimization parameters manually.
```
```suggestion
The Search Service automatically tunes your Vector Search indexes to achieve a balance between: 

* Recall, or the quality of your search results
* Latency, or your search response time
* Memory efficiency

This tuning occurs during indexing and querying. 
You do not need to adjust these parameters manually. 
```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+latency (search response time), and memory efficiency.
+This tuning occurs during indexing and querying, eliminating the need for users to adjust optimization parameters manually.
+
+Couchbase dynamically adjusts two critical vector parameters:
```
```suggestion
Specifically, the Search Service dynamically adjusts two critical vector parameters:
```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+
+Couchbase dynamically adjusts two critical vector parameters:
+
+`nlist` (or `Centroid` count)::
```
```suggestion
`nlist`, also known as `Centroid` count::
```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+
+`nlist` (or `Centroid` count)::
+The number of clusters used for indexing.
+The centroids are used to quickly find the surrounding closest matches in the index. Increasing the number of centroids will increase accuracy but will decrease the speed of the search.
```
```suggestion
Centroids are used to quickly find the surrounding closest matches in the Vector Search index. 
Increasing the number of centroids will increase accuracy but will decrease the speed of the search.
```
Remember to use ventilated prose, please.

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+The number of clusters used for indexing.
+The centroids are used to quickly find the surrounding closest matches in the index. Increasing the number of centroids will increase accuracy but will decrease the speed of the search.
++
+The `nlist` is determined dynamically based on the size of the dataset (number of vectors) in a partition:
```
```suggestion
The `nlist` is determined dynamically based on the size of the dataset, or the number of vectors in a partition:
```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+| `nlist` calculation
+| Notes
+
+| stem:["nvec " ge " 200,000"]
+| stem:[4 xx sqrt("nvec")]
```
Can you build a preview site, please? I want to see how this renders in the table.

RayOffiah replied:
Yup, can do that.

https://preview.docs-test.couchbase.com/doc-12610/server/current/vector-search/fine-tune-vector-search.html

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+
+| stem:["nvec" lt 1000]
+| N/A
+| For number of vectors less than 1000, Couchbase will carry out a straight forward one-to-one mapping between IDs and vectors with an exact vector comparison.
```
```suggestion
| For a number of vectors less than 1000, the Search Service will carry out a straight forward one-to-one mapping between IDs and vectors with an exact vector comparison.
```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+|===
+
+`nprobes` (or `probes`)::
+The number of `centroids` that a query will check for similar vectors.
```
```suggestion
The number of `centroids` that a Search query will check for similar vectors.
```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+
+`nprobes` (or `probes`)::
+The number of `centroids` that a query will check for similar vectors.
+The `nprobe` value is only set when Couchbase is using an Inverted File Index. Couchbase will select the best index type and comparison method depending on the size of the dataset and the `vector_index_optimized_for` setting.
```
```suggestion
The `nprobe` value is only set when the Search Service is using an Inverted File Index. The Search Service will select the best index type and comparison method depending on the size of the dataset and your `vector_index_optimized_for` setting.
```
Please add a link to where we explain the `vector_index_optimized_for` setting. I know it's discussed in creation topics.

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+== Fine-Tuning Query Parameters
+
+
+The Vector Search API will accept an additional parameter object, allowing for tuning the recall and/or accuracy of the search execution.
```
```suggestion
Add an additional parameter object to any Vector Search query to tune the recall or accuracy of your search.
```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+
+The Vector Search API will accept an additional parameter object, allowing for tuning the recall and/or accuracy of the search execution.
+
+`ivf_nprobe_pct`::
+Adjusts the number percentage of clusters searched during queries, allowing for fine-tuning of the balance between recall and performance.
+If the value of `nprobe` is 5% of `nlist` (the centroid count), then setting the value of `ivf_nprobe_pct` higher than 5% will have the search cover a higher percentage of clusters, which will improve the accuracy of the search.
+
+`ivf_max_codes_pct`::
+The default value for this is 100, and the value represents the percentage of `centroids` that will be visited during a search. Reducing the value reduces the number of centroids visited, which will descrease accurary and recal, but will result in faster compute times.
```
```suggestion
You can add the following parameters to your query: 

`ivf_nprobe_pct`::
Adjusts the number percentage of clusters searched during queries, allowing for fine-tuning of the balance between recall and performance.
If the value of `nprobe` is 5% of `nlist` (the centroid count), then setting the value of `ivf_nprobe_pct` higher than 5% will have the search cover a higher percentage of clusters, which will improve the accuracy of the search.

`ivf_max_codes_pct`::
The default value for this is 100, and the value represents the percentage of `centroids` that will be visited during a search. 
Reducing the value reduces the number of centroids visited, which will decrease accuracy and recall, but will result in faster compute times.
```

These need to be added to the full Search documentation, with the explanation of all of the Search request parameters: https://docs.couchbase.com/cloud/search/search-request-params.html

If you don't feel up to doing that yourself, please log a ticket and I will pick it up.

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+`ivf_max_codes_pct`::
+The default value for this is 100, and the value represents the percentage of `centroids` that will be visited during a search. Reducing the value reduces the number of centroids visited, which will descrease accurary and recal, but will result in faster compute times.
+
+For example, given a cluster `vector_index_optimized_for`: "recall" and `indexPartitions`: 5,  `nlist` and `nprobe` are determined based on the current vector count in a given partition:
```
```suggestion
== Example: nlist and nprobe Calculations on a Vector Search Index Set to Recall 

For example, if you had a Vector Search index with `vector_index_optimized_for` set to `"recall"` and `indexPartitions` set to `5`,  `nlist` and `nprobe` are determined based on the current vector count in a given partition:
```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,132 @@
+
+'''
+
+[source, json]
+.Using tuning parameters
+----
+{
+  "fields": ["*"],
+  "knn": [{
+    "k": 10,
+    "params": {
+      "ivf_nprobe_pct": 1,
+      "ivf_max_codes_pct": 0.2
+    },
+    "field": "embedding",
+    "vector": [0.024901132253900747, 1535]
+  }]
+}
+----
```
How does this sample code relate to what's in the table above? The values provided are completely different from what you show in the calculation. What is this doing?

sarahlwelton commented on modules/vector-search/pages/run-vector-search-ui.adoc:
```diff
@@ -7,34 +7,34 @@
 {description}
 
-For more information about how the Search Service scores documents in search results, see xref:run-searches.adoc#scoring[Scoring for Search Queries].
+For more information about how the Search Service scores documents in search results, see xref:fts:fts-scoring.adoc[Scoring for Search Queries].
```
```suggestion
For more information about how the Search Service scores documents in search results, see xref:search:run-searches.adoc#scoring[Scoring for Search Queries].
```

Wrong link. Please don't link to old search doc.

sarahlwelton commented on modules/vector-search/pages/run-vector-search-sdk.adoc:
```diff
@@ -9,57 +9,57 @@
 {description}
 
-For more information about how the Search Service scores documents in search results, see xref:run-searches.adoc#scoring[Scoring for Search Queries].
+For more information about how the Search Service scores documents in search results, see xref:fts:fts-scoring.adoc[Scoring for Search Queries].
```
```suggestion
For more information about how the Search Service scores documents in search results, see xref:search:run-searches.adoc#scoring[Scoring for Search Queries].
```
Wrong link. Please don't link to old search doc.

sarahlwelton reviewed (changes requested):
Lot of noise in this PR with just deleting so many end-of-line spaces. I also have some concerns about how this is shaping up right now. Let me know if you want to discuss.

RayOffiah committed:
[DOC=12610]: Moved fine-tuning menu entry.
[DOC=12610]: Moved the stem declaration; wasn't working in the old position.

RayOffiah committed:
Merge branch 'prerelease/7.6.4' into DOC-12610-Document-`nprobe`-tweaking

RayOffiah committed:
[DOC-12610]: Implemented review changes.

[DOC-12610]:
Moved and reworded the sections for clarity.
Adding link to search-index-params.adoc

sarahlwelton commented on modules/search/pages/search-index-params.adoc:
```diff
@@ -1077,7 +1077,7 @@ The child field's type. Can be one of:
 For more information about the available field data types, see xref:field-data-types-reference.adoc[].
 
-|vector_index_optimized_for |String |Vector Only a|
+|[#vector-index-optimized-param]#vector_index_optimized_for# |String |Vector Only a|
```
Why did we add the multiple hashes around this? I didn't think we needed to do that for table entries.

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,149 @@
+This is the number of `centroids` that a Search query will check for similar vectors.
+The `nprobe` value is only set when the Search Service is using an Inverted File Index. The Search Service will select the best index type and comparison method depending on the size of the dataset and your `vector_index_optimized_for` setting. 
++
+(For more information on the `vector_index_optimized_for` setting, see  xref:search:search-index-params.adoc#vector-index-optimized-param[Search Index JSON properties ])
```
```suggestion
For more information on the `vector_index_optimized_for` setting, see  xref:search:search-index-params.adoc#vector-index-optimized-param[Search Index JSON properties].
```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,149 @@
+==  Default `nlist` and `nprobe` calculations on a Vector Search Index
+
+
+The cluster maintains two dynamically adjusted parameters that will affect the speed, accuracy, and resources used during the search: 
```
```suggestion
The cluster maintains two dynamically adjusted parameters that will affect the speed, accuracy, and resources used during a search: 
```

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,149 @@
+}
+----
+
+In the example above, the search will be carried out across all the fields, looking for close matches for the numbers specified in `vector`. 
```
```suggestion
In the preceding example, the Search request returns results from any available fields in the index, but specifically searches the `embedding` field for the `10` closest matches to the vector `[0.024901132253900747, . . .]`.
The vector in the Search request has been truncated to reduce the display size of the example.
```

sarahlwelton reviewed (commented):
Sorry, I'm being really picky but this is definitely shaping up really well :) Thanks for taking this on, Ray!

sarahlwelton commented on modules/vector-search/pages/fine-tune-vector-search.adoc:
```diff
@@ -0,0 +1,149 @@
+      "ivf_max_codes_pct": 0.2
+    },
+    "field": "embedding",
+    "vector": [0.024901132253900747, 1535]
```
```suggestion
    "vector": [0.024901132253900747, . . .]
```

sarahlwelton reviewed (changes requested):
Sorry, actually saw the example you took this from. It's wrong in its current state. Please add this to the fixes, too.

sarahlwelton reviewed (approved):
Thanks @RayOffiah!
